File: library/src/com/google/maps/android/geojsonkmlabs/Feature.java
package com.google.maps.android.geojsonkmlabs;

import com.google.maps.android.geojsonkmlabs.geojson.GeoJsonLineStringStyle;
import com.google.maps.android.geojsonkmlabs.geojson.GeoJsonPointStyle;
import com.google.maps.android.geojsonkmlabs.geojson.GeoJsonPolygonStyle;

import java.util.HashMap;
import java.util.Observable;

/**
 * Created by suvercha on 12/15/16.
 */

public class Feature extends Observable {
    private final String mId;

    private final HashMap<String, String> mProperties;

    private Geometry mGeometry;

    private GeoJsonPointStyle mPointStyle;

    private GeoJsonLineStringStyle mLineStringStyle;

    private GeoJsonPolygonStyle mPolygonStyle;


    public Feature(Geometry featureGeometry, String id,
                          HashMap<String, String> properties){
        mGeometry = featureGeometry;
        mId = id;
        if (properties == null) {
            mProperties = new HashMap<String, String>();
        } else {
            mProperties = properties;
        }
        //leaving bounding box out for now
    }

    /**
     * Returns all the stored property keys
     *
     * @return iterable of property keys
     */
    public Iterable<String> getPropertyKeys() {
        return mProperties.keySet();
    }

    /**
     * Gets the property entry set
     *
     * @return property entry set
     */
    public Iterable getProperties() {
        return mProperties.entrySet();
    }

    /**
     * Gets the value for a stored property
     *
     * @param property key of the property
     * @return value of the property if its key exists, otherwise null
     */
    public String getProperty(String property) {
        return mProperties.get(property);
    }


    public String getId() {
        return mId;
    }

    /**
     * Checks whether the given property key exists
     *
     * @param property key of the property to check
     * @return true if property key exists, false otherwise
     */
    public boolean hasProperty(String property) {
        return mProperties.containsKey(property);
    }

    /**
     * Gets the geometry object
     *
     * @return geometry object
     */
    public Geometry getGeometry() {
        return mGeometry;
    }

    /**
     * Gets whether the placemark has a properties
     *
     * @return true if there are properties in the properties hashmap, false otherwise
     */
    public boolean hasProperties() {
        return mProperties.size() > 0;
    }

    /**
     * Checks if the geometry is assigned
     *
     * @return true if feature contains geometry object, otherwise null
     */
    public boolean hasGeometry() {
        return (mGeometry != null);
    }


}

File: library/src/main/java/com/google/maps/android/data/Feature.kt
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.google.maps.android.data

import java.util.Observable

/**
 * An abstraction that shares the common properties of
 * [com.google.maps.android.data.kml.KmlPlacemark] and
 * [com.google.maps.android.data.geojson.GeoJsonFeature]
 */
open class Feature(
    geometry: Geometry<*>?,
    id: String?,
    properties: Map<String, String>?
) : Observable() {
    open var id: String? = id
        protected set

    private val _properties: MutableMap<String, String> = properties?.toMutableMap() ?: mutableMapOf()

    open var geometry: Geometry<*>? = geometry
        protected set(value) {
            field = value
            setChanged()
            notifyObservers()
        }

    /**
     * Returns all the stored property keys
     */
    val propertyKeys: Iterable<String>
        get() = _properties.keys

    /**
     * Gets the property entry set
     */
    val properties: Iterable<Map.Entry<String, String>>
        get() = _properties.entries

    /**
     * Gets the value for a stored property
     *
     * @param property key of the property
     * @return value of the property if its key exists, otherwise null
     */
    fun getProperty(property: String): String? = _properties[property]

    /**
     * Checks whether the given property key exists
     *
     * @param property key of the property to check
     * @return true if property key exists, false otherwise
     */
    fun hasProperty(property: String): Boolean = _properties.containsKey(property)

    /**
     * Gets whether the placemark has properties
     *
     * @return true if there are properties in the properties map, false otherwise
     */
    fun hasProperties(): Boolean = _properties.isNotEmpty()

    /**
     * Checks if the geometry is assigned
     *
     * @return true if feature contains geometry object, otherwise null
     */
    fun hasGeometry(): Boolean = geometry != null

    /**
     * Store a new property key and value
     *
     * @param property      key of the property to store
     * @param propertyValue value of the property to store
     * @return previous value with the same key, otherwise null if the key didn't exist
     */
    protected open fun setProperty(property: String, propertyValue: String): String? {
        val prev = _properties.put(property, propertyValue)
        setChanged()
        notifyObservers()
        return prev
    }

    /**
     * Removes a given property
     *
     * @param property key of the property to remove
     * @return value of the removed property or null if there was no corresponding key
     */
    protected open fun removeProperty(property: String): String? {
        val prev = _properties.remove(property)
        setChanged()
        notifyObservers()
        return prev
    }
}
